src/absil/illib.fs

```diff
@@ -523,6 +524,124 @@ module ResultOrException =
+    let bind f (Cancellable f1) = 
+       Cancellable (fun ct -> 
+           if ct.IsCancellationRequested then 
+               ValueOrCancelled.Cancelled (OperationCanceledException()) 
+           else
+               match f1 ct with 
+               | ValueOrCancelled.Value res1 -> 
+                   if ct.IsCancellationRequested then 
+                       ValueOrCancelled.Cancelled (OperationCanceledException()) 
+                   else
+                       let (Cancellable f2) = f res1
+                       f2 ct 
+               | ValueOrCancelled.Cancelled err1 -> 
+                   ValueOrCancelled.Cancelled err1)
+
+    /// Map the result of a cancellable computation
+    let map f (Cancellable f1) = 
+       Cancellable (fun ct -> 
+           match f1 ct with 
+           | ValueOrCancelled.Value res1 -> ValueOrCancelled.Value (f res1)
+           | ValueOrCancelled.Cancelled err1 -> ValueOrCancelled.Cancelled err1)
+                    
+    /// Return a simple value as the result of a cancellable computation
+    let ret x = Cancellable (fun _ -> ValueOrCancelled.Value x)
+
+    /// Fold a cancellable computation along a sequence of inputs
+    let fold f acc seq = 
+        (ret acc,seq) ||> Seq.fold  (fun acc x -> acc |> bind (fun acc -> f acc x))
+    
+    /// Iterate a cancellable computation over a collection
+    let each f seq = 
+        ([],seq) ||> fold (fun acc x -> f x |> map (fun x2 -> x2::acc)) |> map List.rev
+    
+    /// Delay a cancellable computation
+    let delay (f: unit -> Cancellable<'T>) = Cancellable (fun ct -> let (Cancellable g) = f() in g ct)
+
+    /// Run a cancellable computation using the given cancellation token
+    let run ct (Cancellable f) = f ct
+
+    /// Run the computation in a mode where it may not be cancelled. The computation never results in a 
+    /// ValueOrCancelled.Cancelled.
+    let runWithoutCancellation comp = 
+        let res = run System.Threading.CancellationToken.None comp 
+        match res with 
+        | ValueOrCancelled.Cancelled _ -> failwith "unexpected cancellation" 
+        | ValueOrCancelled.Value r -> r
+
+    /// Bind the cancellation token associated with the computation
+    let token () = Cancellable (fun ct -> ValueOrCancelled.Value ct)
+
+    /// Represents a canceled computation
+    let canceled() = Cancellable (fun _ -> ValueOrCancelled.Cancelled (new OperationCanceledException()))
+
+    /// Catch exceptions in a computation
+    let private catch (Cancellable e) = 
+        Cancellable (fun ct -> 
+            try 
+                match e ct with 
+                | ValueOrCancelled.Value r -> ValueOrCancelled.Value (Choice1Of2 r) 
+                | ValueOrCancelled.Cancelled e -> ValueOrCancelled.Cancelled e 
+            with err -> 
+                ValueOrCancelled.Value (Choice2Of2 err))
+
+    /// Implement try/finally for a cancellable computation
+    let tryFinally e compensation =    
```

Comment:
It provides a try/finally semantics for ``cancellable { ... }`` blocks via F# computation expressions (see the ``TryFinally`` and ``Using`` methods on ``CancellableBuilder``.  But as with ``async`` this is not precisely identical to the try/finally semantics of normal, synchronous .NET code.

The semantics is correctly implemented I believe: in the implementation of ``tryFinally`` the body of the computation is run, any exception is reified as data, and regardless of cancellation or not the compensation is executed, then the exception is converted from data back to a result or exception.

See also examples of builders in Expert F# 4.0 or ``eventually`` further below.